gen/field.go

```diff
@@ -467,6 +472,113 @@ func (f fieldGroupGenerator) FromWire(g Generator) error {
 		`, f, TemplateFunc("constantValuePtr", ConstantValuePtr))
 }
 
+func (f fieldGroupGenerator) Decode(g Generator) error {
+	return g.DeclareFromTemplate(
+		`
+		<$stream := import "go.uber.org/thriftrw/protocol/stream">
+
+		<$v := newVar "v">
+		<$sr := newVar "sr">
+		// Decode deserializes a <.Name> struct directly from its Thrift-level
+		// representation, without going through an intemediary type.
+		//
+		// An error is returned if a <.Name> struct could not be generated from the wire
+		// representation.
+		func (<$v> *<.Name>) Decode(<$sr> <$stream>.Reader) error {
+			<$isSet := newNamespace>
+			<range .Fields>
+				<- if .Required ->
+					<$isSet.NewName (printf "%sIsSet" .Name)> := false
+				<- end>
+			<end>
+
+			if err := <$sr>.ReadStructBegin(); err != nil {
+				return err
+			}
+
+			<$fh := newVar "fh">
+			<$ok := newVar "ok">
+			<$fh>, <$ok>, err := <$sr>.ReadFieldBegin()
+			if err != nil {
+				return err
+			}
+
+			for <$ok> {
+				switch <$fh>.ID {
+				<range .Fields ->
+				case <.ID>:
+					if <$fh>.Type == <typeCode .Type> {
+						<- $lhs := printf "%s.%s" $v (goName .) ->
+						<- if .Required ->
+							<$lhs>, err = <decode .Type $sr>
+						<- else ->
+							<decodePtr .Type $lhs $sr>
+						<- end>
+						if err != nil {
+							return err
+						}
+						<if .Required ->
+							<$isSet.Rotate (printf "%sIsSet" .Name)> = true
+						<- end>
+					}
+				<end ->
+				}
+
+				<$fh>, <$ok>, err = <$sr>.ReadFieldBegin()
+				if err != nil {
+					return err
+				}
+			}
+
+			if err := <$sr>.ReadFieldEnd(); err != nil {
```

Comment:
Hmm, shouldn't this be in the for-loop where we're reading the fields? It's a no-op so it doesn't really matter, but for consistency?

Reply:
Ah, yeah, good call.  Moved this into the look before the next `ReadFieldBegin` call.